### PR TITLE
Added the auto completion for components

### LINKED ISSUE #939: Suggestions for component names
parrent: #902 

Suggesting component names:
 -  `odo delete`
 -  `odo component`
 - `odo component set`
 - `odo link`
 -  `odo push`
 -  `odo watch`
 -  + every command that has `--compnent`` flag

### PR DESCRIPTION
**Pull request by mik-dass**
fixes #939

### CODE REVIEW

**Comment by mik-dass** [comment]
@metacosm should I wait till #985 is merged?

**Comment by metacosm** [comment]
@mik-dass up to you, if you get this merged first, I’ll adapt my PR to your code! ;)

**Comment by mik-dass** [comment]
@metacosm Actually I need #985 for showing better suggestions based on the flags (it's not working properly with the current code) :D So I wait

**Comment by mik-dass** [comment]
@metacosm Ready for review

File: pkg/odo/util/completion/completionhandlers.go [hunk 1]
```diff
@@ -79,3 +80,22 @@ var ProjectNameCompletionHandler = func(cmd *cobra.Command, args parsedArgs, con
...
+	completions = make([]string, 0)
+	components, err := component.List(context.Client, context.Application)
+	if err != nil {
+		return completions
+	}
+
+	for _, component := range components {
+		// we found the component name in the list which means
+		// that the component name has been already selected by the user so no need to suggest more
+		if val, ok := args.commands[component.Name]; ok && val {
```
Comment (metacosm) [inline pkg/odo/util/completion/completionhandlers.go]: You only need to do: `if args.commands[component.Name] {` here.

File: cmd/cmdutils.go [hunk 2]
```diff
@@ -157,8 +157,10 @@ func addProjectFlag(cmd *cobra.Command) {
 
 func addComponentFlag(cmd *cobra.Command) {
 	cmd.Flags().String(util.ComponentFlagName, "", "Component, defaults to active component.")
+	completion.RegisterCommandFlagHandler(cmd, "component", completion.ComponentNameCompletionHandler)
 }
 
 func addApplicationFlag(cmd *cobra.Command) {
 	cmd.Flags().String(util.ApplicationFlagName, "", "Application, defaults to active application")
+	completion.RegisterCommandFlagHandler(cmd, "app", completion.AppCompletionHandler)
```
Comment (metacosm) [inline cmd/cmdutils.go]: That's not strictly related to this PR but is appropriate anyway :)

File: cmd/cmdutils.go [hunk 3]
```diff
@@ -157,8 +157,10 @@ func addProjectFlag(cmd *cobra.Command) {
 
 func addComponentFlag(cmd *cobra.Command) {
 	cmd.Flags().String(util.ComponentFlagName, "", "Component, defaults to active component.")
+	completion.RegisterCommandFlagHandler(cmd, "component", completion.ComponentNameCompletionHandler)
```
Comment (metacosm) [inline cmd/cmdutils.go]: 👍

**Comment by mik-dass** [comment]
@metacosm Fixed

**Comment by geoand** [comment]
@mik-dass Could you please rebase this so we can re-test? Thank you

**Comment by mik-dass** [comment]
@geoand ok sure

**Comment by geoand** [comment]
Thanks @mik-dass !

**Comment by geoand** [comment]
@mik-dass Cool :)

**Comment by metacosm** [comment]
> @geoand Actually I need some inputs from @metacosm regarding the new changes before I rebase. I don't want to break anything accidentally :)

What do you need, @mik-dass?

**Comment by mik-dass** [comment]
> What do you need, @mik-dass?

I want to remove the addProjectFlag() functions calls and call the genericclioptions stuff directly. But am I missing something?

File: pkg/odo/util/completion/completionhandlers.go [hunk 4]
```diff
@@ -75,7 +76,7 @@ var ProjectNameCompletionHandler = func(cmd *cobra.Command, args parsedArgs, con
 	for _, project := range projects {
 		// we found the project name in the list which means
 		// that the project name has been already selected by the user so no need to suggest more
-		if val, ok := args.commands[project.Name]; ok && val {
+		if args.commands[project.Name] {
```
Comment (geoand) [inline pkg/odo/util/completion/completionhandlers.go]: Is this guaranteed to work?

If so, can you please apply it to the rest of the handlers as well?

File: pkg/odo/util/completion/completion.go [hunk 5]
```diff
@@ -159,3 +159,18 @@ func getUserTypedCommands(args complete.Args, command *cobra.Command) []string {
 
 	return commands
 }
+
+func AddProjectFlag(cmd *cobra.Command) {
```
Comment (geoand) [inline pkg/odo/util/completion/completion.go]: Can you add some comments about this and the following functions as well?
Them being in the completion package makes it a little tricky to immediately understand that they actually add a cobra flag without looking at the source code :)

File: pkg/odo/util/completion/completionhandlers.go (hunk 4 above)
Comment (metacosm) [reply pkg/odo/util/completion/completionhandlers.go]: This works if the map is of type `map[string]bool`, which it is :)

Comment (geoand) [reply pkg/odo/util/completion/completionhandlers.go]: Yeah, just found a relevant SO answer :)

However this change should also be applied to the other functions in this file as well

File: pkg/odo/util/completion/completion.go (hunk 5 above)
Comment (geoand) [reply pkg/odo/util/completion/completion.go]: I agree. However the solution seems better than the alternative :)

**Comment by metacosm** [comment]
I would like to get more opinions on this new re-organization before merging.

**Comment by metacosm** [comment]
So @kadel, @cdrage: what should we do with respect to the design issue?

**Comment by kadel** [comment]
> @metacosm @kadel How about I put those functions in some cmdUtil file?

not sure about that, would that help?


How about adding them to respective cli packages?
`AddProjectFlag` to `github.com/redhat-/developer/odo/pkg/odo/cli/project`
`AddComponentFlag` to `github.com/redhat-/developer/odo/pkg/odo/cli/component`
`AddApplicationFlag` to `github.com/redhat-/developer/odo/pkg/odo/cli/application`

**Comment by mik-dass** [comment]
> How about adding them to respective cli packages?
AddProjectFlag to github.com/redhat-/developer/odo/pkg/odo/cli/project
AddComponentFlag to github.com/redhat-/developer/odo/pkg/odo/cli/component
AddApplicationFlag to github.com/redhat-/developer/odo/pkg/odo/cli/application

Good idea! Done Please have a look :)

**Comment by metacosm** [comment]
@mik-dass @kadel I don't think this new version is a good idea either because this adds even more links between packages that shouldn't be related but if you're fine with it, I can live with it as well :)

**Comment by mik-dass** [comment]
> @mik-dass @kadel I don't think this new version is a good idea either because this adds even more links between packages that shouldn't be related but if you're fine with it, I can live with it as well :)

I agree it's still a bad design but at least it is better than putting those functions in the completion.go file. So let's leave it for now, we can refactor it later on, WDYT? :)

**Comment by geoand** [comment]
Let's just rebase the PR first :P

**Pull request merged**